File: xdm-core/src/main/java/xdm/core/network/http/HttpResponse.java
package xdm.core.network.http;

import java.io.InputStream;
import java.net.URI;
import java.time.LocalDateTime;

public interface HttpResponse extends AutoCloseable {
  void close();

  int getStatusCode();

  String getStatusMessage();

  long getContentLength();

  String getContentDisposition();

  String getContentType();

  LocalDateTime getLastModified();

  InputStream getInputStream();

  boolean isRedirected();

  URI getFinalUrl();

  String getHeader(String name);
}

File: xdm-core/src/main/java/xdm/core/network/http/HttpResponse.kt
package xdm.core.network.http

import java.io.InputStream
import java.net.URI
import java.time.LocalDateTime

interface HttpResponse : AutoCloseable {
    override fun close()
    val statusCode: Int
    val statusMessage: String
    val contentLength: Long?
    val contentDisposition: String?
    val contentType: String?
    val lastModified: LocalDateTime?
    val inputStream: InputStream
    val isRedirected: Boolean
    val finalUrl: String?
    fun getHeader(name: String): String?
}

class HttpResponseImpl(
    override val contentDisposition: String?,
    override val statusCode: Int,
    override val statusMessage: String,
    override val contentLength: Long?,
    override val contentType: String?,
    override val lastModified: LocalDateTime?,
    override val inputStream: InputStream,
    override val isRedirected: Boolean,
    override val finalUrl: String?,
    private val closeCallback: () -> Unit,
    private val headerCallback: (String) -> String?,
) : HttpResponse {
    override fun close() {
        closeCallback()
    }

    override fun getHeader(name: String): String? = headerCallback(name)
}
